packages/core/src/tools/ToolsManager.ts

```diff
@@ -1,27 +1,211 @@
+import 'reflect-metadata';
+import { deepMerge, isFunction, isObject, PromiseQueue } from '@editorjs/helpers';
+import { Inject, Service } from 'typedi';
+import type { BlockToolConstructor } from "../entities/BlockTool.js";
+import {
+  BlockToolBuilder, BlockTuneBuilder,
+  InlineToolBuilder,
+  ToolsCollection,
+  ToolsFactory,
+  UnifiedToolConfig
+} from "./builders/index.js";
 import { Paragraph } from './internal/paragraph/index.js';
-import type { EditorConfig } from '@editorjs/editorjs';
+import type {
+  EditorConfig,
+  InlineTool,
+  ToolConfig,
+  ToolConstructable,
+  ToolSettings
+} from '@editorjs/editorjs';
 
 /**
  * Works with tools
+ * @todo - validate tools configurations
+ * @todo - merge internal tools
  */
+@Service()
 export default class ToolsManager {
+  #factory: ToolsFactory;
+  #config: UnifiedToolConfig;
+  #availableTools = new ToolsCollection();
+  #unavailableTools = new ToolsCollection();
+
+  /**
+   * Returns available Tools
+   */
+  public get available(): ToolsCollection {
+    return this.#availableTools;
+  }
+
+  /**
+   * Returns unavailable Tools
+   */
+  public get unavailable(): ToolsCollection {
+    return this.#unavailableTools;
+  }
+
+  /**
+   * Return Tools for the Inline Toolbar
+   */
+  public get inlineTools(): ToolsCollection<InlineToolBuilder> {
+    return this.available.inlineTools;
+  }
+
+  /**
+   * Return editor block tools
+   */
+  public get blockTools(): ToolsCollection<BlockToolBuilder> {
+    return this.available.blockTools;
+  }
+
+  /**
+   * Return available Block Tunes
+   * @returns - object of Inline Tool's classes
+   */
+  public get blockTunes(): ToolsCollection<BlockTuneBuilder> {
+    return this.available.blockTunes;
+  }
+
+  /**
+   * Returns internal tools
+   */
+  public get internal(): ToolsCollection {
+    return this.available.internalTools;
+  }
+
+  /**
+   * @param editorConfig - EditorConfig object
+   * @param editorConfig.tools - Tools configuration passed by user
+   */
+  constructor(@Inject('EditorConfig') editorConfig: EditorConfig) {
+    this.#config = this.#prepareConfig(editorConfig.tools ?? {});
+
+    this.#validateTools();
+
+    this.#factory = new ToolsFactory(this.#config, editorConfig, {});
+
+    void this.#prepareTools(editorConfig.tools);
+  }
+
+  #prepareTools(config: ToolConfig): Promise<void> {
+    const promiseQueue = new PromiseQueue();
+
+    Object.entries(this.#config).forEach(([toolName, config]) => {
+      if (isFunction(config.class.prepare)) {
+        promiseQueue.add(async () => {
+          try {
+            await config.class.prepare!({
+              toolName: toolName,
+              config: config,
+            });
+
+            const tool = this.#factory.get(toolName);
+
+            if (tool.isInline()) {
+              /**
+               * Some Tools validation
+               */
+              const inlineToolRequiredMethods = ['render'];
```

Comment:
maybe move this constant to the sdk package?

Reply:
We can do that later